notificationdemo/source/java/src/org/kaaproject/kaa/demo/notification/NotificationDemo.java

```diff
@@ -51,74 +55,114 @@ public static void main(String[] args) {
+
+        Scanner scanner = new Scanner(System.in);
+        while (scanner.hasNextLong()) {
+            long topicId = scanner.nextLong();
+            if (getTopic(topicId) != null) {
+                LOG.info("Subscribing to optional topic {}", topicId);
+                subscribeTopic(topicId);
+            } else {
+                LOG.info("There is no input topic id. Please, input existing topic id.");
+            }
         }
 
         // Stop listening to the notification topic list updates.
         kaaClient.removeTopicListListener(topicListListener);
+        unsubscribeOptionalTopics();
 
         // Stop the Kaa client and release all the resources which were in use.
         kaaClient.stop();
         LOG.info("Notification demo stopped");
     }
 
-    // A listener that tracks the notification topic list updates
-    // and subscribes the Kaa client to every new topic available.
-    private static class BasicNotificationTopicListListener implements NotificationTopicListListener {
-        @Override
-        public void onListUpdated(List<Topic> list) {
-            LOG.info("Topic list was updated");
-            showTopicList(list);
-            try {
-                //Try to subscribe to all new optional topics, if any.
-                List<Long> optionalTopics = extractOptionalTopicIds(list);
-                for(Long optionalTopicId : optionalTopics){
-                    LOG.info("Subscribing to optional topic {}", optionalTopicId);
-                }
-                kaaClient.subscribeToTopics(optionalTopics, true);
-            } catch (UnavailableTopicException e) {
-                LOG.error("Topic is unavailable, can't subscribe: {}", e.getMessage());
-            }
+    private static void inputTopicIdMessage() {
+        LOG.info("\nPlease, type topic ID in order to subscribe to ones or type any text to exit: \n");
+    }
+
+    private static void showTopics() {
+        if (topics == null || topics.isEmpty()) {
+            LOG.info("Topic list is empty");
+            return;
         }
+
+        LOG.info("Available topics:");
+        for (Topic topic : topics) {
+            LOG.info("Topic id: {}, name: {}, type: {}", topic.getId(), topic.getName(), topic.getSubscriptionType());
+        }
+
+        LOG.info("Subscribed on topics:");
+        for (Topic t : getOneTypeTopics(SubscriptionType.MANDATORY_SUBSCRIPTION)) {
+            LOG.info("Topic id: {}, name: {}, type: {}", t.getId(), t.getName(), t.getSubscriptionType().name());
+        }
+        inputTopicIdMessage();
     }
 
-    private static List<Long> extractOptionalTopicIds(List<Topic> list) {
-        List<Long> topicIds = new ArrayList<>();
-        for (Topic t : list) {
-            if (t.getSubscriptionType() == SubscriptionType.OPTIONAL_SUBSCRIPTION) {
-                topicIds.add(t.getId());
+    private static List<Topic> getOneTypeTopics(SubscriptionType type) {
+        List<Topic> topics = new ArrayList<>();
```

Comment:
Don't cover property [topics] name, you can use something like: [oneTypeTopics] for the local variable name, it will help to avoid [NotificationDemo.topics] construction.